From 13e77f04b01b5ebcfde0f63dd5e9a9922567916e Mon Sep 17 00:00:00 2001
From: Greensi7 <adam.tabak04@gmail.com>
Date: Thu, 10 Sep 2026 18:19:44 +0200
Subject: [PATCH] Fix inconsistent state on LHASH insertion failure

If lh_CONF_VALUE_insert() fails in _CONF_add_string() then new
CONF value is still left in stack. This can prevent
proper duplicate field replacement causing multiple values
with same field name to stay on stack.

Assisted-by: ChatGPT:gpt-5.6
Reviewed-by: Milan Broz <mbroz@openssl.org>
Reviewed-by: Nikola Pajkovsky <nikolap@openssl.org>
Merge-date: Tue Oct 27 08:53:48 2026
Merged-from: https://github.com/openssl/openssl/pull/33037
---
 crypto/conf/conf_api.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/crypto/conf/conf_api.c b/crypto/conf/conf_api.c
index c615d007..ebdcb94c 100644
--- a/crypto/conf/conf_api.c
+++ b/crypto/conf/conf_api.c
@@ -54,6 +54,11 @@ int _CONF_add_string(CONF *conf, CONF_VALUE *section, CONF_VALUE *value)
         return 0;
 
     v = lh_CONF_VALUE_insert(conf->data, value);
+    if (lh_CONF_VALUE_error(conf->data)) {
+        (void)sk_CONF_VALUE_pop(ts);
+        return 0;
+    }
+
     if (v != NULL) {
         (void)sk_CONF_VALUE_delete_ptr(ts, v);
         OPENSSL_free(v->name);
